File: testgmres.cpp
#include"gmres.h"

//=========================================================
void findAX(double *AX,double *X,int n)
{
	for(int i=0;i<n;i++)
	{
		if(i == 0)
		{
			AX[i] = -2*X[i]+X[i+1];
		}
		else
		if( i == n-1)
		{
			AX[i]=-2*X[i]+X[i-1];
		}
		else
		{
			AX[i]=X[i+1]+X[i-1]-2*X[i];
		}
	}
}
//=========================================================
void noprecond(double *MinvX,double *X,int n)
{
	for(int i=0;i<n;i++)
	{
		MinvX[i]=X[i];
	}
}
//=========================================================
void gauss_seidel(double *MinvX,double *X,int n)
{
	double cf1,cf2,cf3;
	double left,right;
	int numit;

	cf1=1.0; cf2=-2.0; cf3=1.0;
	numit=500;

	for(int it=0;it<numit;it++)
	{
		for(int i=0;i<n;i++)
		{
			left   = (i==0)?0.0:MinvX[i-1];
			right  = (i==n-1)?0.0:MinvX[i+1];

			MinvX[i] = (X[i] -cf1*left -cf3*right)/cf2;
		}
	}
}
//=========================================================
void findbvec(double *b,int n)
{
	b[0]   = 0.0;
	b[n-1] = -1.0;
}
//=========================================================
int main(int argc,char *argv[])
{
	double *x0,*x,*b;
	int m,n,it;
	double *testvec;
	std::string pctype;
	std::ofstream outfile("soln.dat");

	if(argc == 2)
	{
		pctype=argv[1];
	}
	else
	{
		pctype="none";
	}

	m=40; n=50; it=20;

	x0 = new double[n]();
	b  = new double[n]();
	x  = new double[n]();

	solvergmres obj;

	findbvec(b,n);

	obj.setkspvalues(m,n,it);

	if(pctype == "gs")
	{
		obj.performgmres(b,x0,x,&findAX,&gauss_seidel);
	}
	else
	{
	 obj.performgmres(b,x0,x,&findAX,&noprecond);
	}

	//gauss_seidel(x,b,n);

	for(int i=0;i<n;i++)
	{
		outfile<<i<<"\t"<<x[i]<<"\n";
	}

	return(0);
}
//=========================================================

File: testgmres.f90
!=========================================================================
subroutine noprecond(MinvX,X,n)

      implicit none

      integer,intent(in) :: n
      real*8, intent(in) :: X(n)
      real*8,intent(out) :: MinvX(n)

      MinvX=X

end subroutine noprecond
!=========================================================================
subroutine findAX(AX,X,n)
      
	implicit none

	integer,intent(in) :: n
        real*8, intent(in) :: X(n)
        real*8,intent(out) :: AX(n)

	real*8 :: sup,sub,diag
	integer :: i

	sup  =  1.0
	sub  =  1.0
	diag = -2.0

	AX = 0.d0

	do i=1,n

		if(i .eq. 1) then
			AX(i)=diag*x(i)+sup*x(i+1)
		else if(i .eq. n) then
			AX(i)=diag*x(i)+sub*x(i-1)
		else
			AX(i)=sub*x(i-1)+diag*x(i)+sup*x(i+1)
		endif
	enddo

end subroutine findAX
!=========================================================================
subroutine findbvec(b,n)
	
	integer, intent(in) :: n
	real*8, intent(out)  :: b(n)

	b=0.d0

	b(1) = 0.d0
	b(n) =-1.d0

end subroutine findbvec
!=========================================================================
program testgmres
      use solvergmres
      implicit none

      integer :: m,n,it,i
      real*8,allocatable :: x0(:),x(:),b(:)
      external :: findAX,noprecond

      open(unit=5,file="soln.dat")

      m=40
      n=50
      it=20

      allocate(x0(n))
      allocate(x(n))
      allocate(b(n))

      x0 = 0.d0
      x  = 0.d0

      call findbvec(b,n)

      call performgmres(b,x0,x,m,n,it,findAX,noprecond)

      do i=1,n
      	write(5,'(I5 F10.5)'),i,x(i)
      enddo

      close(5)

end program testgmres
!=========================================================================
